modules/configuration/src/main/java/org/apache/ignite/configuration/internal/NamedListConfiguration.java

```diff
@@ -0,0 +1,125 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.configuration.internal;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.BiFunction;
+import java.util.stream.Collectors;
+import org.apache.ignite.configuration.internal.property.Modifier;
+import org.apache.ignite.configuration.internal.property.NamedList;
+
+/**
+ * TODO: Add class description.
+ *
+ * @author @java.author
+ * @version @java.version
+ */
+public class NamedListConfiguration<U, T extends Modifier<U, INIT, CHANGE>, INIT, CHANGE> extends DynamicConfiguration<NamedList<U>, NamedList<INIT>, NamedList<CHANGE>> {
+    /** Creator of named configuration. */
+    private final BiFunction<String, String, T> creator;
+
+    /** Named configurations. */
+    Map<String, T> values = new HashMap<>();
+
+    /**
+     * Constructor.
+     * @param prefix
+     * @param key
+     * @param configurator
+     * @param root
+     * @param creator
+     */
+    public NamedListConfiguration(
+        String prefix,
+        String key,
+        Configurator<? extends DynamicConfiguration<?, ?, ?>> configurator,
+        DynamicConfiguration<?, ?, ?> root,
+        BiFunction<String, String, T> creator
+    ) {
+        super(prefix, key, false, configurator, root);
+        this.creator = creator;
+    }
+
+    /**
+     * Constructor.
+     * @param base
+     * @param configurator
+     * @param root
+     */
+    private NamedListConfiguration(
+        NamedListConfiguration<U, T, INIT, CHANGE> base,
+        Configurator<? extends DynamicConfiguration<?, ?, ?>> configurator,
+        DynamicConfiguration<?, ?, ?> root
+    ) {
+        super(base.prefix, base.key, false, configurator, root);
+
+        this.creator = base.creator;
+
+        base.values.forEach((key, value) -> {
+            final T copy = (T) ((DynamicConfiguration<U, INIT, CHANGE>) value).copy(root);
```

Comment:
oof, this is rough. Something's wrong. Do we have to have "T extends Modifier<U, INIT, CHANGE>" instead of just "Modifier<U, INIT, CHANGE>". This might be a stupid question because I don't understand meaning of these parameters :)